cmd/gpu_plugin/gpu_plugin_resource_manager.go

```diff
@@ -0,0 +1,506 @@
+			return nil, errRetry
+		}
+		candidate = timestampedCandidates[0]
+	}
+
+	candidate.pod = pendingPods[candidate.name]
+	return &candidate, nil
+}
+
+// +kubebuilder:rbac:groups="",resources=pods,verbs=list
+
+// getNodePendingGPUPods returns a map of pod names -> pods that are pending and use the gpu.
+func (rm *resourceManager) getNodePendingGPUPods() (map[string]*v1.Pod, error) {
+	selector, err := fields.ParseSelector("spec.nodeName=" + rm.nodeName +
+		",status.phase=" + string(v1.PodPending))
+
+	if err != nil {
+		klog.Error(err.Error())
+		return nil, err
+	}
+
+	pendingPodList, err := rm.clientset.CoreV1().Pods(v1.NamespaceAll).List(context.TODO(), metav1.ListOptions{
+		FieldSelector: selector.String(),
+	})
+
+	if err != nil {
+		klog.Error(err.Error())
+		return nil, err
+	}
+
+	// make a map ouf of the list, accept only GPU-using pods
+	pendingPods := make(map[string]*v1.Pod)
+	for i := range pendingPodList.Items {
+		pod := &pendingPodList.Items[i]
+
+		if numGPUUsingContainers(pod) > 0 {
+			pendingPods[pod.Name] = pod
+		}
+	}
+
+	return pendingPods, nil
+}
+
+// findAllocationPodCandidates returns a slice of all potential allocation candidate pods.
+// This goes through the PODs listed in the podresources grpc service and finds those among pending
+// pods which don't have all GPU devices allocated.
+func (rm *resourceManager) findAllocationPodCandidates(pendingPods map[string]*v1.Pod) (candidates []podCandidate, err error) {
+	candidates = []podCandidate{}
+
+	resListerClient, clientConn, err := podresources.GetV1Client(grpcAddress, grpcTimeout, grpcBufferSize)
+	if err != nil {
+		return nil, errors.Wrap(err, "Could not get a grpc client for reading plugin resources")
+	}
+
+	defer clientConn.Close()
+	ctx, cancel := context.WithTimeout(context.Background(), grpcTimeout)
+	defer cancel()
+
+	resp, err := resListerClient.List(ctx, &podresourcesv1.ListPodResourcesRequest{})
+	if err != nil {
+		return nil, errors.Wrap(err, "Could not read plugin resources via grpc")
+	}
+
+	for _, podRes := range resp.PodResources {
+		// count allocated gpu-using containers
+		numContainersAllocated := 0
+		for _, cont := range podRes.Containers {
+			for _, dev := range cont.Devices {
+				if dev.ResourceName == resourcePrefix+deviceType {
+					numContainersAllocated++
+					break
+				}
+			}
+		}
+
+		if pod, pending := pendingPods[podRes.Name]; pending {
+			allocationTargetNum := numGPUUsingContainers(pod)
+			if numContainersAllocated < allocationTargetNum {
+				candidate := podCandidate{
+					pod:                 pod,
+					name:                podRes.Name,
+					allocatedContainers: numContainersAllocated,
+					allocationTargetNum: allocationTargetNum,
+				}
+				candidates = append(candidates, candidate)
+			}
+		}
+	}
+
+	return candidates, nil
+}
+
+func (rm *resourceManager) Scan(devTree dpapi.DeviceTree) {
+	rm.mutex.Lock()
+	defer rm.mutex.Unlock()
+	rm.devTree = devTree
+}
```

Comment:
This is basically setting `resourceManager`'s `devTree` field. Let's rename `Scan()` to `setDevTree()`?

Reply:
sounds good to me, done